README.md
# TestDimorph

<!-- badges: start -->
[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/TestDimorph)](https://cran.r-project.org/package=TestDimorph)
[![metacran downloads](https://cranlogs.r-pkg.org/badges/grand-total/TestDimorph)](https://cran.r-project.org/package=TestDimorph)
[![Codecov test coverage](https://codecov.io/gh/bassam-abulnoor/TestDimorph/branch/master/graph/badge.svg)](https://codecov.io/gh/bassam-abulnoor/TestDimorph?branch=master)
<!-- badges: end -->

Analysis Of The Interpopulation Difference In Degree of Sexual Dimorphism Using Summary Statistics

## Installation

You can install the released version of TestDimorph from [CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("TestDimorph")
```
or the development version from GitHub:

``` r
devtools::install_github("bassam-abulnoor/TestDimorph")
```

## Data entry

The input can be in the form of wide format **raw data** or **summary statistics** where each row represents measurements of a single individual or population. In order to be recognized by the functions, columns need to have specific names (case sensitive):


1. **Sex** : either **M** for male or **F** for female (*Raw data*)
2. **Pop** : studied populations (*Summary/Raw data*)
3. **Trait** : studied traits or parameters (*Summary*)
4. **m** : male sample size (*Summary*)
5. **f** : female sample size (*Summary*)
6. **M.mu**: male mean (*Summary*)
7. **F.mu** : female mean (*Summary*)
8. **M.sdev** : male standard deviation (*Summary*)
9. **F.sdev** : female standard deviation (*Summary*)

**N.B**

* All non numerical values need to be factors.
* The list and the data frame formats for the `multivariate` and `RawGen` functions input, can be be found in the attached data sets under the names `baboon.parms_list` and `baboon.parms_df` respectively.

## Examples of data entry

#### Summary statistics for univariate analysis

##### Comparison of maximal femur length and mediolateral diameter in mid-shaft


|    Pop    |  m  | M.mu  | M.sdev |  f  | F.mu  | F.sdev |
| :-------: | :-: | :---: | :----: | :-: | :---: | :----: |
|  Turkish  | 150 | 49.39 |  3.01  | 150 | 42.91 |  2.90  |
| Bulgarian | 82  | 48.33 |  2.53  | 58  | 42.89 |  2.84  |



#### Summary statistics for multivariate analysis and raw data generation

##### Comparison of maximal femur length and mediolateral diameter in mid-shaft


|    Pop    | Parms |  m  |  M.mu  | M.sdev |  f  |  F.mu  | F.sdev |
| :-------: | :---: | :-: | :----: | :----: | :-: | :----: | :----: |
| Bulgarian | MXFL  | 82  | 461.80 | 19.90  | 58  | 411.70 | 23.20  |
|   Greek   | MXFL  | 36  | 440.40 | 19.60  | 34  | 409.80 | 21.40  |
| Bulgarian |  MLD  | 82  | 27.67  |  2.21  | 58  | 24.89  |  1.78  |
|   Greek   |  MLD  | 36  | 27.74  |  1.79  | 34  | 26.69  |  2.42  |



#### Raw data for summary extraction

##### The Howells' craniometric data

| Sex |  Pop  | GOL | NOL | BNL | BBH | XCB | XFB | ZYB | AUB |
| :-: | :---: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
|  M  | NORSE | 189 | 185 | 100 | 135 | 143 | 120 | 133 | 119 |
|  F  | NORSE | 179 | 177 | 97  | 127 | 140 | 114 | 122 | 118 |
|  M  | EGYPT | 192 | 190 | 104 | 132 | 140 | 116 | 134 | 125 |
|  F  | EGYPT | 174 | 172 | 93  | 123 | 140 | 118 | 121 | 114 |


#### Summary statistics for van_vark function

##### Howells summary data

| Trait |  Pop  |   M.mu   |  m |   F.mu   |  f |
|:-----:|:-----:|:--------:|:--:|:--------:|:--:|
|  GOL  | EGYPT | 185.6207 | 58 | 175.5849 | 53 |
|  GOL  | NORSE | 188.4727 | 55 | 179.9818 | 55 |
|  NOL  | EGYPT |  183.569 | 58 | 175.0377 | 53 |
|  NOL  | NORSE | 186.1818 | 55 | 178.6909 | 55 |


NEWS.md
# TestDimorph 0.4.0

* A major update.
* New functions:`van_vark`, `Hedges_g`,`MI_index` and `D_index`.
* `univariate` and `multivariate` functions report different types of ANOVA and MANOVA.
* effect sizes are reported with confidence intervals.
* added 3 new datasets.
* added a list `models` of the supported models for `accu_model` function.

# TestDimorph 0.3.6

* Minor fixes-no user visible changes.

# TestDimorph 0.3.5

* Maintainer email changed.

# TestDimorph 0.3.3

* The package has been re-written and optimized with fewer dependencies.
* New function names with snake case letters are introduced
* Function names written in camel case letters are deprecated
* `accu_model` function now can do cross-validation using different models supported by `caret` package.

# TestDimorph 0.3.1

* `mda` is removed from dependencies and mixture and flexible discriminant analyses are no longer options in the `AccuModel` function.

# TestDimorph 0.3.0

* Fixed some issues with corrplots in `Tg` function.

# TestDimorph 0.2.9

* Raw data generation now can be either by uni or multivariate lognormal or truncated distribution.
* Added the package to github.
* Output of most functions is in the form of tidy tibbles.
* Effect size for uni and multivariate analyses.
* `pMatrix` function is deprecated and included in `Tg` function. 
* `AccuModel` function can generate roc curves.
* added random forest to methods of `AccuModel`
* `aovSS` function can do different types of post hoc tests.
* Pairwise comparisons can be expressed by means of different alphabetical letters.
* Informative error messages in case of wrong type of input.

# TestDimorph 0.2.1

* Minor fixes-no user visible changes.

# TestDimorph 0.2.0

* Added a `NEWS.md` file to track changes to the package.
* Removed `biotools` from the dependencies to allow the package to work on Mac OS X.
* Added 2 new functions `RawGen` and `AccuModel`.
* Now `pMatrix` function can generate corrplots.
* Added a `README.md`file.
* updated the the package's description and help files.

 


R/aov_ss.R
#' @title Sex Specific One way ANOVA From Summary statistics
#' @description Calculates sex specific one way ANOVA from summary statistics.
#' @inheritParams univariate
#' @inheritParams t_greene
#' @param pairwise Logical; if TRUE runs multiple pairwise comparisons on
#' different populations using Tukey's post hoc test, Default: TRUE
#' @return Sex specific ANOVA tables and pairwise comparisons in tidy format.
#' @details Data is entered as a data frame of summary statistics where
#' the column containing population names is chosen by position (first by
#' default), other columns of summary data should have specific names (case
#' sensitive) similar to \link{baboon.parms_df}
#' @examples
#' # Comparisons of femur head diameter in four populations
#' library(TestDimorph)
#' df <- data.frame(
#'   Pop = c("Turkish", "Bulgarian", "Greek", "Portuguese "),
#'   m = c(150.00, 82.00, 36.00, 34.00),
#'   f = c(150.00, 58.00, 34.00, 24.00),
#'   M.mu = c(49.39, 48.33, 46.99, 45.20),
#'   F.mu = c(42.91, 42.89, 42.44, 40.90),
#'   M.sdev = c(3.01, 2.53, 2.47, 2.00),
#'   F.sdev = c(2.90, 2.84, 2.26, 2.90)
#' )
#' aov_ss(x = df)
#' @rdname aov_ss
#' @export
#' @importFrom stats rnorm aov TukeyHSD
#' @importFrom multcompView multcompLetters
aov_ss <-
  function(x,
           Pop = 1,
           pairwise = TRUE,
           letters = FALSE,
           es_anova = "none",
           digits = 4,
           CI = 0.95) {
    # Data preparation --------------------------------------------------------

    if (!(is.data.frame(x))) {
      stop("x should be a dataframe")
    }
    if (!all(c("M.mu", "F.mu", "M.sdev", "F.sdev", "m", "f") %in% names(x))) {
      stop(
        "colnames should contain:
            M.mu= Male mean
            F.mu=Female mean
            M.sdev=Male sd
            F.sdev=Female sd
            m= Male sample size
            f=Female sample size
            N.B: colnames are case sensitive"
      )
    }
    if (!(Pop %in% seq_along(x))) {
      stop("Pop should be number from 1 to ncol(x)")
    }
    if (!is.logical(pairwise)) {
      stop("pairwise should be either TRUE or FALSE")
    }
    if (!is.logical(letters)) {
      stop("letters should be either TRUE or FALSE")
    }
    es_anova <-
      match.arg(es_anova, choices = c("none", "eta", "f"))
    if (CI < 0 ||
      CI > 1 || !is.numeric(CI)) {
      stop("CI should be a number between 0 and 1")
    }
    if(isTRUE(pairwise)&&nlevels(x$Pop)==2){

      warning("Tukey post hoc test was not performed because there are only 2 populations ")
    }
    x <- data.frame(x)
    x$Pop <- x[, Pop]
    x$Pop <-
      gsub(
        x = x$Pop,
        pattern = "[^[:alnum:]]",
        replacement = "_"
      )
    x$Pop <- factor(x$Pop)
    levels(x$Pop) <- droplevels(x$Pop)
    if (length(unique(x$Pop)) != length(x$Pop[!is.na(x$Pop)])) {
      stop("Populations names'should be unique")
    }

    # ANOVA from summary data -------------------------------------------------

    aov_sum <- function(.mu, .sdev, n) {
      N <- length(.mu)
      Pop <- factor(rep(levels(x$Pop), n))
      df <- lapply(seq_len(N), function(i) {
        scale(stats::rnorm(n[i])) * .sdev[i] + .mu[i]
      })
      x <- do.call(rbind, df)
      out <- data.frame(Pop, x)
      return(out)
    }
    av_M <- aov_sum(x$M.mu, x$M.sdev, x$m)
    av_F <- aov_sum(x$F.mu, x$F.sdev, x$f)
    av_M <- stats::aov(x ~ Pop, data = av_M)
    av_F <- stats::aov(x ~ Pop, data = av_F)
    M_model <-
      anova_es(x = av_M, es_anova = es_anova, digits = digits, CI = CI)
    F_model <-
      anova_es(x = av_F, es_anova = es_anova, digits = digits, CI = CI)

    # Pairwise comparisons ----------------------------------------------------
    M_post <-data.frame(TukeyHSD(av_M, conf.level = CI)[[1]])
    M_post <- M_post %>% mutate(across(1:ncol(.),round,digits))
    M_post <-rown_col(M_post,var = "populations")
    colnames(M_post) <-
      c(
        "populations",
        "mean.diff",
        "conf.low",
        "conf.high",
        "p.value"
      )
    p <- M_post$p.value
    M_post$signif <-
      case_when(
        p > 0.05 ~ "ns",
        p < 0.05 & p > 0.01 ~ "*",
        p < 0.01 & p > 0.001 ~ "**",
        p < 0.001 ~ "***"
      )

    M_letters <- M_post$p.value
    names(M_letters) <- M_post$populations
    M_letters <-
      rown_col(data.frame(
        "letters" = multcompView::multcompLetters(M_letters, threshold = CI)[[1]]
      ),
      var = "populations"
      )
    F_post <-
      data.frame(
        TukeyHSD(av_F, conf.level = CI)[[1]]
      )
    F_post <- F_post %>% mutate(across(1:ncol(.),round,digits))
    F_post <-rown_col(F_post,var = "populations")
    colnames(F_post) <-
      c(
        "populations",
        "mean.diff",
        "conf.low",
        "conf.high",
        "p.value"
      )
    p <- F_post$p.value
    F_post$signif <-
      case_when(
        p > 0.05 ~ "ns",
        p < 0.05 & p > 0.01 ~ "*",
        p < 0.01 & p > 0.001 ~ "**",
        p < 0.001 ~ "***"
      )

    F_letters <- F_post$p.value
    names(F_letters) <- F_post$populations
    F_letters <-
      rown_col(data.frame(
        "letters" = multcompView::multcompLetters(F_letters, threshold = CI)[[1]]
      ),
      var = "populations"
      )


    if (isTRUE(pairwise)&& nlevels(x$Pop)>2) {
      if (isTRUE(letters)) {
        list(
          "Male model" = M_model,
          "Male posthoc" = M_post,
          "Male letters" = M_letters,
          "Female model" = F_model,
          "Female posthoc" = F_post,
          "Female letters" = F_letters
        )
      } else {
        list(
          "Male model" = M_model,
          "Male posthoc" = M_post,
          "Female model" = F_model,
          "Female posthoc" = F_post
        )
      }
    } else {
      list("Male model" = M_model, "Female model" = F_model)
    }
  }
